File: src/main/java/net/weavemc/loader/api/command/CommandBus.java
package net.weavemc.loader.api.command;

import lombok.experimental.UtilityClass;
import net.weavemc.loader.api.ModInitializer;
import net.weavemc.loader.api.event.ChatSentEvent;
import net.weavemc.loader.api.event.EventBus;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/**
 * The Command Bus manages commands by Weave mods.
 */
@SuppressWarnings("unused")
@UtilityClass
public class CommandBus {

    private final List<Command> commands = new ArrayList<>();

    /**
     * Commands can be registered at any time during the client lifecycle, but if you intend
     * on normal behavior, you should register during the {@link ModInitializer#preInit()} phase.
     *
     * @param command The command instance to register.
     */
    public void register(Command command) {
        commands.add(command);
    }

    static {
        EventBus.subscribe(ChatSentEvent.class, e -> {
            if (!e.getMessage().startsWith("/")) return;

            String[] split = e.getMessage().substring(1).split("\\s+");

            Iterator<Command> matching = commands.stream().filter(c -> c.matches(split[0])).iterator();
            if (!matching.hasNext()) return;

            e.setCancelled(true);
            String[] args = Arrays.copyOfRange(split, 1, split.length);
            matching.forEachRemaining(c -> c.handle(args));
        });
    }

}

File: api/v1_8/src/main/kotlin/net/weavemc/api/command/CommandBus.kt
package net.weavemc.api.command

import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiChat
import net.minecraft.network.play.client.C14PacketTabComplete
import net.minecraft.network.play.server.S3APacketTabComplete
import net.weavemc.api.ChatSentEvent
import net.weavemc.api.PacketEvent
import net.weavemc.api.event.EventBus
import net.weavemc.api.event.SubscribeEvent

/**
 * The Command Bus manages commands by Weave mods.
 *
 * This implementation differs from Weave 0.x in that only the first matched command is executed.
 * If multiple commands share a name or aliases, the command registered earliest takes precedence.
 */
object CommandBus {
    private val commands = mutableListOf<Command>()

    private val WHITESPACE_REGEX = Regex("\\s+")

    private fun String.split(): List<String> = split(WHITESPACE_REGEX)

    init {
        arrayOf(
            ChatListener,
            TabCompletionListener
        ).forEach(EventBus::subscribe)
    }

    /**
     * Commands can be registered at any time during the client lifecycle.
     *
     * @param command The command instance to register.
     */
    @JvmStatic
    fun register(command: Command) {
        commands.add(command)
    }

    private object ChatListener {
        @SubscribeEvent
        fun onChatSentEvent(event: ChatSentEvent) {
            val message = event.message.trim()

            if (message[0] != '/') {
                return
            }

            val args = message.drop(1).split()

            commands
                .find { it.matches(args[0]) }
                ?.execute(args.toTypedArray())
        }
    }

    private object TabCompletionListener {
        @Volatile
        var lastCommandAndSuggestions: Pair<Command?, Array<String>>? = null

        private val guiChat: GuiChat?
            get() = Minecraft.getMinecraft().currentScreen as? GuiChat

        @SubscribeEvent
        fun onPacketSend(event: PacketEvent.Send) {
            val packet = event.packet
            if (packet !is C14PacketTabComplete) {
                return
            }

            val message = packet.message.trim()
            val commandArgs = message.split()

            if (commandArgs.size == 1 && message.startsWith('/')) {
                val suggestions = commands
                    .filter(Command::showInRoot)
                    .flatMap { it.matching(message) }
                    .toTypedArray()

                lastCommandAndSuggestions = null to suggestions
            } else {
                val targetBlock = packet.targetBlock

                for (command in commands) {
                    val suggestions = command.getSuggestions(commandArgs.drop(1).toTypedArray(), targetBlock) ?: continue

                    if (command.exclusiveSuggestions) {
                        guiChat?.run {
                            lastCommandAndSuggestions = null

                            onAutocompleteResponse(suggestions)

                            event.cancelled = true
                        }
                    } else {
                        lastCommandAndSuggestions = command to suggestions
                    }

                    break
                }
            }
        }

        @SubscribeEvent
        fun onPacketReceive(event: PacketEvent.Receive) {
            val packet = event.packet
            if (packet !is S3APacketTabComplete) {
                return
            }

            val (lastCommand, lastSuggestions) = lastCommandAndSuggestions ?: return
            val serverSuggestions = packet.matches

            lastCommandAndSuggestions = null

            val orderedSuggestions = lastCommand?.orderSuggestions(lastSuggestions, serverSuggestions)
                ?: (lastSuggestions + serverSuggestions).distinct().sorted().toTypedArray()

            packet.matches = orderedSuggestions
        }
    }
}